### ISSUE TITLE
Some problems about code implementation and paper description

### ISSUE THREAD

**Issue by TianQi-777** [TonghanWang/ROMA#9]
Hello, I found that some implementations are inconsistent with the paper description, I hope you can help me to understand.

### 1. multivariate Gaussian distribution(you can click [here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L96) to check)
In the paper, the role distribution is assumed to be a multivariate Gaussian distribution. But I found that your implementation uses a Gaussian distribution function(NOT multivariate), . In training phase, The shape of the sampling result is `latent = gaussian_embed.rsample()` => `torch.Size([96, 3])` => `torch.Size([batch_size * n_agents, dim])`,  I guess it's because dim=3, which means that each agent's role is represented by a multivariate Gaussian distribution with dim=3, so you needn't use `torch.distributions.MultivariateNormal`

### 2. Regular terms about Identifiable and Versatile Roles([here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L113))
In the derivation of section 3 of your paper, the final loss (Eq. (3)) is equal to CE-H(CrossEntropy-Entropy). In my understanding, this is the KL divergence of q and p, since KL = CE - H, I am confused about the implementation below
```
loss = gaussian_embed.entropy().sum(dim=-1).mean() * self.args.h_loss_weight + \
            kl_divergence(gaussian_embed, gaussian_infer).sum(dim=-1).mean() * self.args.kl_loss_weight   # CE = H + KL
```
If I understand according to your comments, where is -H in the codes(Eq. (3) in the paper)? I prefer to treat `gaussian_embed.entropy().sum(dim=-1).mean() ` in the code as entropy exploration(like TRPO?).

### 3. Softplus operation([here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L116))
Why use softplus here.

### 4. Calculation of mi(Regular terms about Specialized Roles)([here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L133))
In my opinion, `mi` is equivalent to the variation function q in the paper Eq. (6). However, q is `self.latent_infer` in the code, since the input of `self.latent_infer` is `cat[h_in, inputs]`. Therefore, I think the code should be modified as follows

Orig:
```
latent_dis = latent.clone().view(self.bs, self.n_agents, -1)
latent_move = latent.clone().view(self.bs, self.n_agents, -1)

mi = th.clamp(gaussian_embed.log_prob(latent_move.view(self.bs * self.n_agents, -1))+13.9,\
                                  min=-13.9).sum(dim=1,keepdim=True) / self.latent_dim
```
modified:
```
# change latent to latent_infer
latent_dis = latent_infer.clone().view(self.bs, self.n_agents, -1)      
latent_move = latent_infer.clone().view(self.bs, self.n_agents, -1)

# change gaussian_embed to gaussian_infer
mi = th.clamp(gaussian_infer.log_prob(latent_move.view(self.bs * self.n_agents, -1))+13.9,\
                                  min=-13.9).sum(dim=1,keepdim=True) / self.latent_dim
```
### 5. dissimilarity model input problem([here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L129))
In the Eq. (6) of your paper, the input of dissimilarity model is the trajectory of agents i and j. Howerver, in your code, the input of `self.dis_net` is role latent `latent_dis and latent_move`, not `h_in`(trajectory). I am confused about this.

### 6. A small question about the calculation graph in pytorch([here](https://github.com/TonghanWang/ROMA/blob/a5b0d4dd4ac6b6cfefc97bb6f21ccf397f1d525d/src/modules/agents/latent_ce_dis_rnn_agent.py#L107))
Why use `.detach()`here to remove `h_in` from the calculation graph?

### 7. run time
I try to run the code with the command as follows
```
python3 src/main.py --config=qmix_smac_latent --env-config=sc2 with agent=latent_ce_dis_rnn env_args.map_name=3m t_max=20050000
```
However, the maps of `3m` are running very slowly, but I only need 2 days to train `qmix` in `pymarl`, I am confused about this. The result below is that it has been training for 2, 3 days ...
Details:
```
[INFO 22:51:55] my_main Updated target network
[INFO 23:03:18] my_main t_env: 2755241 / 20050000
[INFO 23:03:18] my_main Estimated time left: 15 days, 14 hours, 23 minutes, 39 seconds. Time passed: 2 days, 9 hours, 19 minutes, 13 seconds
[INFO 23:03:22] my_main Recent Stats | t_env:    2755241 | Episode:    43728
battle_won_mean:           0.5831       di_max:                    1.0000       di_mean:                   0.1212       di_min:                    0.0000
di_std:                    0.2519       ep_length_mean:           72.5449       epsilon:                   0.0500       grad_norm:                 1.3739
loss:                      0.7765       loss_ce:                   0.6931       loss_dis:                 -0.0004       loss_reg:                  0.6927
mi_max:                    1.0000       mi_mean:                   0.4992       mi_min:                    0.0000       mi_std:                    0.3931
q_taken_mean:              1.4689       return_mean:              16.6189       return_std:                4.0020       target_mean:               1.4601
td_error_abs:              0.1858       test_battle_won_mean:      0.7417       test_ep_length_mean:      67.6500       test_return_mean:         17.4870
test_return_std:           3.6892       var_mean:                  0.0429
[INFO 23:11:36] my_main Updated target network
[INFO 23:31:16] my_main Updated target network
[INFO 23:32:01] my_main t_env: 2775428 / 20050000
[INFO 23:32:01] my_main Estimated time left: 17 days, 1 hours, 32 minutes, 58 seconds. Time passed: 2 days, 9 hours, 47 minutes, 56 seconds
[INFO 23:32:05] my_main Recent Stats | t_env:    2775428 | Episode:    44016
battle_won_mean:           0.5618       di_max:                    1.0000       di_mean:                   0.1315       di_min:                    0.0000
di_std:                    0.2536       ep_length_mean:           71.9337       epsilon:                   0.0500       grad_norm:                 1.2976
loss:                      0.7732       loss_ce:                   0.6931       loss_dis:                 -0.0004       loss_reg:                  0.6927
mi_max:                    1.0000       mi_mean:                   0.4741       mi_min:                    0.0000       mi_std:                    0.3906
q_taken_mean:              1.4347       return_mean:              16.4456       return_std:                4.0802       target_mean:               1.4272
td_error_abs:              0.1820       test_battle_won_mean:      0.7000       test_ep_length_mean:      69.2000       test_return_mean:         17.2025
test_return_std:           3.8033       var_mean:                  0.0475
[INFO 23:48:52] my_main Updated target network
[INFO 23:58:20] my_main t_env: 2795641 / 20050000
[INFO 23:58:20] my_main Estimated time left: 15 days, 14 hours, 24 minutes, 45 seconds. Time passed: 2 days, 10 hours, 14 minutes, 15 seconds
[INFO 23:58:24] my_main Recent Stats | t_env:    2795641 | Episode:    44304
battle_won_mean:           0.5604       di_max:                    1.0000       di_mean:                   0.1252       di_min:                    0.0000
di_std:                    0.2526       ep_length_mean:           71.1094       epsilon:                   0.0500       grad_norm:                 1.0757
loss:                      0.7700       loss_ce:                   0.6931       loss_dis:                 -0.0004       loss_reg:                  0.6927
mi_max:                    1.0000       mi_mean:                   0.4877       mi_min:                    0.0000       mi_std:                    0.3905
q_taken_mean:              1.4241       return_mean:              16.4276       return_std:                4.1351       target_mean:               1.4185
td_error_abs:              0.1827       test_battle_won_mean:      0.7250       test_ep_length_mean:      68.3583       test_return_mean:         17.3875
test_return_std:           3.6658       var_mean:                  0.0471
[INFO 00:08:17] my_main Updated target network
[INFO 00:25:01] my_main t_env: 2815938 / 20050000
[INFO 00:25:01] my_main Estimated time left: 15 days, 17 hours, 23 minutes, 53 seconds. Time passed: 2 days, 10 hours, 40 minutes, 55 seconds
[INFO 00:25:05] my_main Recent Stats | t_env:    2815938 | Episode:    44584
battle_won_mean:           0.5818       di_max:                    1.0000       di_mean:                   0.1199       di_min:                    0.0000
di_std:                    0.2494       ep_length_mean:           71.3123       epsilon:                   0.0500       grad_norm:                 1.2917
loss:                      0.7714       loss_ce:                   0.6931       loss_dis:                 -0.0004       loss_reg:                  0.6926
mi_max:                    1.0000       mi_mean:                   0.5028       mi_min:                    0.0000       mi_std:                    0.3850
q_taken_mean:              1.3729       return_mean:              16.5951       return_std:                4.0498       target_mean:               1.3659
td_error_abs:              0.1851       test_battle_won_mean:      0.7333       test_ep_length_mean:      68.3750       test_return_mean:         17.4525
test_return_std:           3.6723       var_mean:                  0.0465
```

Looking forward to your reply :)

**Comment by TonghanWang** [comment]
Hi, we are very sorry for the late reply. We were trying to update ROMA to match the newest version of PyMARL. Since some of your questions will be resolved then, we did not reply to your issues on time. (Btw, an updated version of ROMA will be released in a few weeks.)

Q1: We did not use `torch.distributions.MultivariateNormal`:
A1: This is because we use diagonal covariance matrices to reduce the number of parameters. `torch.distributions.Normal` matches the descriptions in our paper, and we find it works well in practice.

Q2: Regular terms about Identifiable and Versatile Roles:
A2: Yes, you are right, in our implementation, we use KL for CE-H. `gaussian_embed.entropy().sum(dim=-1).mean()` is for exploration, and since it is a commonly used trick, we did not describe it in our paper. We will add some discussion about this in the camera-ready version of ROMA. :)

Q3: Softplus operation
A3: This is a trick that we find can make training more stable.

Q4: Calculation of mi
A4: Yes, we realize this problem, and it is supposed to be `gaussian_infer.log_prob`. However, I guess, empirically, these two will have similar performance. Because, intuitively, both of them encourage the roles of different agents to be the same, which is in line with our aim stated in the paper. We will fix this in the next version of this repo.

Q5: Dissimilarity model input
A5: Since we optimize the mutual information between the history and the role, we expect that they are identifiable by each other. Thus, here, we use roles as compact representations of history, which can effectively reduce the number of parameters.

Q6: Use of `detach()`
A6: We did not want the role learning component and the *perception* component to influence each other. This is because the learning of roles is already very difficult; if it can influence the representations of history, the training may be slower.

Q7: Run time
A7: When we carried out our experiments, we found that ROMA has comparable run time with QMIX. A possible reason is that we use 8 parallel environments to collect data. Which `runner` are you using?

Please let us know if our answers do not resolve your concerns. :)

**Comment by TianQi-777** [comment]
Thanks for your reply, most of my doubts have been resolved. Although neural network optimization itself is a black box(some operations or regular terms cannot be explained well), the motivation of this paper is interesting :)
